ios/LayoutReanimation/REAAnimationsManager.m

```diff
@@ -288,73 +242,131 @@ - (NSDictionary *)prepareDataForLayoutAnimatingWorklet:(NSMutableDictionary *)cu
   return preparedData;
 }
 
-- (void)onViewRemoval:(UIView *)view before:(REASnapshot *)before
+- (BOOL)wantsHandleRemovalOfView:(UIView *)view
 {
-  NSNumber *tag = view.reactTag;
-  ViewState state = [_states[tag] intValue];
-  if (state == Disappearing || state == ToRemove || tag == nil) {
-    return;
+  return REANodeFind(view, ^(id<RCTComponent> view) {
+    return [self->_ancestorsOfExitingViews objectForKey:view.reactTag] != nil ||
+        self->_hasAnimationForTag(view.reactTag, @"exiting");
+  });
+}
+
+- (void)registerExitingAncestors:(UIView *)child
+{
+  UIView *parent = child.superview;
+  while (parent != nil && ![parent isKindOfClass:[RCTRootView class]]) {
+    if (parent.reactTag != nil) {
+      _ancestorsOfExitingViews[parent.reactTag] = @([_ancestorsOfExitingViews[parent.reactTag] intValue] + 1);
+    }
+    parent = parent.superview;
   }
-  NSMutableDictionary<NSString *, NSObject *> *startValues = before.values;
-  if (state == Inactive) {
-    if (startValues != nil) {
-      _states[tag] = [NSNumber numberWithInt:ToRemove];
-      [_toRemove addObject:tag];
-      [self scheduleCleaning];
+}
+
+- (void)maybeDropAncestors:(UIView *)child
+{
+  UIView *parent = child.superview;
+  while (parent != nil && ![parent isKindOfClass:[RCTRootView class]]) {
+    UIView *view = parent;
+    parent = parent.superview;
+    if (view.reactTag != nil) {
+      int trackingCount = [_ancestorsOfExitingViews[view.reactTag] intValue] - 1;
+      if (trackingCount <= 0) {
+        if ([_ancestorsToRemove containsObject:view.reactTag]) {
+          [view removeFromSuperview];
+          [_ancestorsToRemove removeObject:view.reactTag];
+        }
+        [_ancestorsOfExitingViews removeObjectForKey:view.reactTag];
+      } else {
+        _ancestorsOfExitingViews[view.reactTag] = @(trackingCount);
+      }
     }
-    return;
   }
-  _states[tag] = [NSNumber numberWithInt:Disappearing];
-  NSDictionary *preparedValues = [self prepareDataForAnimatingWorklet:startValues frameConfig:ExitingFrame];
-  _startAnimationForTag(tag, @"exiting", preparedValues, @(0));
 }
 
-- (void)onViewCreate:(UIView *)view after:(REASnapshot *)after
+- (BOOL)removeRecursive:(UIView *)view fromContainer:(UIView *)container withoutAnimation:(BOOL)shouldRemove;
 {
-  _reaUiManager.flushUiOperations();
-  NSNumber *tag = view.reactTag;
-  if (_states[tag] == nil) {
-    _states[tag] = [NSNumber numberWithInt:Inactive];
-    _viewForTag[tag] = view;
+  if (view.reactTag) {
+    BOOL hasExitAnimation = _hasAnimationForTag(view.reactTag, @"exiting");
+    BOOL wantAnimateExit = hasExitAnimation;
+
+    for (UIView *subview in [view.reactSubviews copy]) {
+      if ([self removeRecursive:subview fromContainer:view withoutAnimation:(shouldRemove && !hasExitAnimation)]) {
+        wantAnimateExit = YES;
```

Comment:
First I thought we can `break;` the loop here but `[self removeRecursive]` has side-effects so now I think we cannot.